cmd/juju/application/deployer/charm.go

```diff
@@ -286,17 +286,16 @@ func (d *predeployedLocalCharm) PrepareAndDeploy(ctx *cmd.Context, deployAPI Dep
 		return errors.Trace(err)
 	}
 	formattedCharmURL := d.userCharmURL.String()
-	ctx.Infof("Located charm %q.", formattedCharmURL)
-	ctx.Infof("Deploying charm %q.", formattedCharmURL)
-
-	d.origin, err = deduceOrigin(userCharmURL)
-	if err != nil {
-		return errors.Trace(err)
-	}
 
 	d.id = charmstore.CharmID{URL: d.userCharmURL}
 	d.series = userCharmURL.Series
+	d.origin, err = utils.DeduceOrigin(userCharmURL, "")
+	if err != nil {
+		return err
+	}
 
+	ctx.Infof("Located charm %q.", formattedCharmURL)
+	ctx.Infof("Deploying charm %q.", formattedCharmURL)
```

Comment:
Off-topic, I am not a fan of this, it could easily be summarised into one. There is zero code happening between them both and as a user, you think it's doing something, that it's not.

Reply:
spread out info appropriately